drivers/usb/device/usb_dc_nrf5.c

```diff
@@ -0,0 +1,2256 @@
+		__ASSERT_NO_MSG(ep_ctx);
+
+		ret = k_mem_pool_alloc(&ep_buf_pool, &ep_ctx->buf.block,
+				       MAX_ISO_EP_BUF_SZ, K_NO_WAIT);
+		if (ret < 0) {
+			SYS_LOG_ERR("EP buffer alloc failed for ISOIN");
+			__ASSERT_NO_MSG(0);
+		}
+
+		ep_ctx->buf.data = ep_ctx->buf.block.data;
+		ep_ctx->buf.curr = ep_ctx->buf.data;
+	}
+
+	if (CFG_EP_ISOOUT_CNT) {
+		ep_ctx = nrf5_usbd_out_endpoint_ctx(NRF_USBD_EPOUT(8));
+		__ASSERT_NO_MSG(ep_ctx);
+
+		ret = k_mem_pool_alloc(&ep_buf_pool, &ep_ctx->buf.block,
+				       MAX_ISO_EP_BUF_SZ, K_NO_WAIT);
+		if (ret < 0) {
+			SYS_LOG_ERR("EP buffer alloc failed for ISOOUT");
+			__ASSERT_NO_MSG(0);
+		}
+
+		ep_ctx->buf.data = ep_ctx->buf.block.data;
+		ep_ctx->buf.curr = ep_ctx->buf.data;
+	}
+}
+
+static void nrf5_usbd_endpoint_ctx_deinit(void)
+{
+	struct nrf5_usbd_ep_ctx *ep_ctx;
+	u32_t i;
+
+	for (i = 0; i < CFG_EPIN_CNT; i++) {
+		ep_ctx = nrf5_usbd_in_endpoint_ctx(i);
+		__ASSERT_NO_MSG(ep_ctx);
+		k_mem_pool_free(&ep_ctx->buf.block);
+		memset(ep_ctx, 0, sizeof(*ep_ctx));
+	}
+
+	for (i = 0; i < CFG_EPOUT_CNT; i++) {
+		ep_ctx = nrf5_usbd_out_endpoint_ctx(i);
+		__ASSERT_NO_MSG(ep_ctx);
+		k_mem_pool_free(&ep_ctx->buf.block);
+		memset(ep_ctx, 0, sizeof(*ep_ctx));
+	}
+
+	if (CFG_EP_ISOIN_CNT) {
+		ep_ctx = nrf5_usbd_in_endpoint_ctx(NRF_USBD_EPIN(8));
+		__ASSERT_NO_MSG(ep_ctx);
+		k_mem_pool_free(&ep_ctx->buf.block);
+		memset(ep_ctx, 0, sizeof(*ep_ctx));
+	}
+
+	if (CFG_EP_ISOOUT_CNT) {
+		ep_ctx = nrf5_usbd_out_endpoint_ctx(NRF_USBD_EPOUT(8));
+		__ASSERT_NO_MSG(ep_ctx);
+		k_mem_pool_free(&ep_ctx->buf.block);
+		memset(ep_ctx, 0, sizeof(*ep_ctx));
+	}
+}
+
+static int nrf5_usbd_power_int_enable(bool enable)
+{
+	struct device *dev = device_get_binding(CONFIG_USBD_NRF5_NAME);
+
+	if (!dev) {
+		SYS_LOG_ERR("could not get USBD power device binding");
+		return -ENODEV;
+	}
+
+	nrf5_power_usb_power_int_enable(dev, enable);
+
+	return 0;
+}
+
+int usb_dc_attach(void)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+	int ret;
+
+	if (ctx->attached) {
+		ret = 0;
+		goto err_attached;
+	}
+
+	k_work_init(&ctx->usb_work, nrf5_usbd_work);
+	k_fifo_init(&ctx->work_queue);
+	k_sem_init(&ctx->dma_in_use, 1, 1);
+
+	ret = nrf5_usbd_power_int_enable(true);
+	if (ret) {
+		goto err_pwr_int_enable;
+	}
+
+	nrf5_usbd_install_isr();
+	nrf5_usbd_enable_interrupts();
+
+	ret = nrf5_usbd_hf_clock_enable(true, false);
+	if (ret) {
+		goto err_clk_enable;
+	}
+
+	nrf5_usbd_endpoint_ctx_init();
+	ctx->attached = true;
+
+	return ret;
+
+err_clk_enable:
+	nrf5_usbd_disable_interrupts();
+	nrf5_usbd_power_int_enable(false);
+
+err_pwr_int_enable:
+err_attached:
+
+	return ret;
+}
+
+int usb_dc_detach(void)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+	int ret;
+
+	nrf5_usbd_disable_interrupts();
+
+	if (nrf_usbd_pullup_check()) {
+		nrf_usbd_pullup_disable();
+	}
+
+	nrf_usbd_disable();
+
+	ret = nrf5_usbd_hf_clock_enable(false, false);
+	if (ret) {
+		goto err_clk_disable;
```

Comment:
`return ret` here directly, no need for `goto`

Reply:
Ack